File: multilive/platforms/yy.py
"""YY 直播。

抓取策略：
  列表：频道页纯 HTTP（SSR 页面内嵌全部在播房间，平台房间不多，
    一次抓全；已实测 music=28 / dancing=21 / pretty=100 左右）。
  播放地址：YY 官方播放器 SDK 走 stream-manager.yy.com/v3/channel/streams
    （带 SDK 签名/指纹，纯 HTTP 复现返回 result:2 被拒），最终
    *-flv-web.yy.com 直链只能在浏览器网络层捕获。参考 douyu_warm：
    浏览器打开房间页抓 FLV 直链（tools/yy_live.mjs），缓存到
    output/yy_warm.json，后续刷新直接复用未过期直链，只对新增/过期房间
    重新开浏览器（房间少，单轮约几分钟）。

m3u 语义：keep_stale=False，只保留在播房间。
"""
import json
import os
import re
import shutil
import subprocess
import time
from concurrent.futures import ThreadPoolExecutor

from multilive.config import Source
from multilive.core import Room, Session, fmt_exc, log

NAME = 'yy'
keep_stale = False

BASE = 'https://www.yy.com/{}'
WARM_SCRIPT = 'tools/yy_live.mjs'
CACHE_NAME = 'yy_warm.json'
WARM_TTL = 30 * 60        # 直链缓存 30 分钟，超时重新开浏览器取流
WARM_TIMEOUT = 900        # 整体取流超时（房间多时放宽）
LIST_WORKERS = 3

GROUP_MAP = {
    'dancing': '舞蹈',
    'pretty': '颜值',
    'music': '音乐',
    'sing': '音乐',
}


def parse(line):
    m = re.match(r'https?://www\.yy\.com/([a-zA-Z0-9_]+)/?$', line.strip())
    if m:
        src = Source(NAME, 'category', m.group(1))
        return [src]
    return []


def fetch_category(sess, cat):
    """SSR 页面 -> {sid: {sid, ssid, title, nickname, group}}。"""
    rooms = {}
    try:
        _st, html, _ = sess.get_text(BASE.format(cat),
                                     referer='https://www.yy.com/')
    except Exception as e:
        log().info('  [频道] %s 页面失败: %s', cat, fmt_exc(e))
        return rooms
    gm = re.search(r'data-stat-name="([^"]+)"', html)
    group = GROUP_MAP.get(cat) or (gm.group(1).strip() if gm else cat)
    for block in re.split(r'<li\b', html):
        m = re.search(r'data-url="/(\d+)/(\d+)\?(?:[^"]*?tempId=\d+[^"]*?)?"'
                      r'[^>]*?data-title="([^"]*)"', block)
        if not m:
            continue
        sid, ssid, title = m.groups()
        title = (title or '').strip()
        nm = re.search(r'<span class="intro">([^<]*)</span>', block)
        nick = (nm.group(1) if nm else '').strip()
        am = re.search(r'data-original="([^"]+)"', block)
        avatar = (am.group(1) if am else '').strip()
        if not sid:
            continue
        rooms[sid] = {
            'sid': sid, 'ssid': ssid or sid,
            'title': title, 'nickname': nick or title,
            'group': group,
            'avatar': avatar,
        }
    log().info('  [频道] %s: %d 个在播房间', cat, len(rooms))
    return rooms


def cache_path(ctx):
    return os.path.join(ctx.project_root, 'output', CACHE_NAME)


def load_cache(ctx):
    try:
        with open(cache_path(ctx), encoding='utf-8') as f:
            return json.load(f) or {}
    except Exception:
        return {}


def save_cache(ctx, cache):
    try:
        os.makedirs(os.path.dirname(cache_path(ctx)), exist_ok=True)
        with open(cache_path(ctx), 'w', encoding='utf-8') as f:
            json.dump(cache, f, ensure_ascii=False)
    except Exception as e:
        log().debug('缓存 yy_warm 失败: %s', fmt_exc(e))


def warm_rooms(ctx, rooms):
    """对新增/过期房间浏览器取流；返回 {sid: flv直链}。"""
    cache = load_cache(ctx)
    now = time.time()
    need = []
    for sid, meta in rooms.items():
        ent = cache.get(sid) or {}
        if not ent.get('url') or now - float(ent.get('ts') or 0) > WARM_TTL:
            need.append({'sid': sid, 'ssid': meta.get('ssid') or sid})
    if need:
        script = os.path.join(ctx.project_root, WARM_SCRIPT)
        if not (os.path.exists(script) and shutil.which('node')):
            log().warning('[yy] 无 %s 或 Node，本轮 %d 个房间跳过取流',
                          WARM_SCRIPT, len(need))
            return {}
        payload = json.dumps({'rooms': need, 'concurrency': 3})
        try:
            r = subprocess.run(['node', script], input=payload,
                               capture_output=True, text=True,
                               timeout=WARM_TIMEOUT)
            if r.returncode != 0:
                raise RuntimeError((r.stderr.strip() or r.stdout.strip())[-200:])
            got = json.loads(r.stdout).get('urls') or {}
            fresh = 0
            for sid, url in got.items():
                if url:
                    cache[sid] = {'url': url, 'ts': now}
                    fresh += 1
            save_cache(ctx, cache)
            log().info('  [取流] 浏览器取 %d/%d 个房间直链（缓存 %d 个）',
                       fresh, len(need), len(cache))
        except Exception as e:
            log().warning('[yy] 浏览器取流失败(%s)，沿用未过期的缓存直链',
                          fmt_exc(e))
    out = {}
    for sid in rooms:
        ent = cache.get(sid) or {}
        if ent.get('url'):
            out[sid] = ent['url']
    return out


def fetch(sources, ctx):
    log().info('[yy] %d 个来源', len(sources))
    cats = sorted({s.target for s in sources if s.kind == 'category'})
    rooms = {}
    with ThreadPoolExecutor(max_workers=LIST_WORKERS) as ex:
        results = ex.map(lambda c: fetch_category(Session(), c), cats)
        for cat_rooms in results:
            for sid, meta in cat_rooms.items():
                if sid not in rooms:
                    rooms[sid] = meta
    log().info('  [列表] 共 %d 个不重复房间', len(rooms))
    if not rooms:
        return []
    urls = warm_rooms(ctx, rooms)
    out = []
    for sid, meta in rooms.items():
        url = urls.get(sid)
        if not url:
            continue
        out.append(Room(
            platform=NAME, rid=sid,
            title=meta['title'], nickname=meta['nickname'],
            url=url, group=meta['group'] or NAME,
            avatar=meta.get('avatar') or ''))
    log().info('[yy] 完成: %d 在播房间', len(out))
    return out

File: internal/platforms/yy/yy.go
// Package yy YY 直播：频道页 SSR 第一页 + /more/page.action 分页补齐。
package yy

import (
	"fmt"
	"regexp"
	"sort"
	"strings"

	"multilive/internal/core"
	"multilive/internal/platform"
)

const playerBase = "https://astar.cc.cd/yy/%s"
const catAPI = "https://www.yy.com/more/page.action"
const pageSize = 200

var (
	catRe   = regexp.MustCompile(`^https?://www\.yy\.com/([a-zA-Z0-9_]+)/?$`)
	liRe    = regexp.MustCompile(`data-url="/(\d+)/(\d+)\?[^"]*?"[^>]*?data-title="([^"]*)"`)
	nickRe  = regexp.MustCompile(`<span class="intro">([^<]*)</span>`)
	avRe    = regexp.MustCompile(`data-original="([^"]+)"`)
	statRe  = regexp.MustCompile(`data-stat-name="([^"]+)"`)
	barRe   = regexp.MustCompile(`pageBar\s*:\s*\{([^}]*)\}`)
	numRe   = func(key string) *regexp.Regexp { return regexp.MustCompile(key + `\s*:\s*(\d+)`) }
	strRe   = func(key string) *regexp.Regexp { return regexp.MustCompile(key + `\s*:\s*'([^']*)'`) }
	titleRe = regexp.MustCompile(`\s*正在直播$`)
)

var groupMap = map[string]string{
	"dancing": "舞蹈", "pretty": "颜值", "music": "音乐", "sing": "音乐",
}

type meta struct {
	sid, ssid, title, nickname, group, avatar string
}

type Platform struct{ platform.Base }

func (Platform) Name() string    { return "yy" }
func (Platform) KeepStale() bool { return false }

func (Platform) Parse(line string) ([]core.Source, error) {
	if m := catRe.FindStringSubmatch(strings.TrimSpace(line)); m != nil {
		return []core.Source{{Platform: "yy", Kind: "category", Target: m[1]}}, nil
	}
	return nil, nil
}

func grabNum(bar, key string) int {
	m := numRe(key).FindStringSubmatch(bar)
	if m == nil {
		return 0
	}
	n := 0
	fmt.Sscanf(m[1], "%d", &n)
	return n
}

func grabStr(bar, key string) string {
	m := strRe(key).FindStringSubmatch(bar)
	if m == nil {
		return ""
	}
	return m[1]
}

func fetchCategory(cat string) map[string]meta {
	rooms := map[string]meta{}
	sess := core.NewClient(nil)
	_, body, _, err := sess.GetText("https://www.yy.com/"+cat, "https://www.yy.com/")
	if err != nil {
		core.Infof("  [频道] %s 页面失败: %s", cat, core.FmtErr(err, 120))
		return rooms
	}
	group := groupMap[cat]
	if group == "" {
		if m := statRe.FindStringSubmatch(body); m != nil {
			group = strings.TrimSpace(m[1])
		}
		if group == "" {
			group = cat
		}
	}
	for _, block := range strings.Split(body, "<li") {
		m := liRe.FindStringSubmatch(block)
		if m == nil {
			continue
		}
		sid, ssid, title := m[1], m[2], titleRe.ReplaceAllString(strings.TrimSpace(m[3]), "")
		nick := ""
		if nm := nickRe.FindStringSubmatch(block); nm != nil {
			nick = strings.TrimSpace(nm[1])
		}
		avatar := ""
		if am := avRe.FindStringSubmatch(block); am != nil {
			avatar = strings.TrimSpace(am[1])
		}
		if nick == "" {
			nick = title
		}
		rooms[sid] = meta{sid: sid, ssid: ssid, title: title, nickname: nick, group: group, avatar: avatar}
	}
	if pm := barRe.FindStringSubmatch(body); pm != nil {
		bar := pm[1]
		moduleID := grabNum(bar, "moduleId")
		total := grabNum(bar, "totalCount")
		biz := grabStr(bar, "biz")
		sub := grabStr(bar, "subBiz")
		if biz != "" && moduleID != 0 && total != 0 {
			pages := (total + pageSize - 1) / pageSize
			for page := 1; page <= pages; page++ {
				api := fmt.Sprintf("%s?biz=%s&subBiz=%s&page=%d&moduleId=%d&pageSize=%d",
					catAPI, biz, sub, page, moduleID, pageSize)
				var j map[string]any
				_, _, err := sess.GetJSON(api, "https://www.yy.com/"+cat, &j, 0)
				if err != nil {
					core.Infof("  [频道] %s 第%d页失败: %s", cat, page, core.FmtErr(err, 120))
					continue
				}
				var items []any
				if dm := core.JMap(j, "data"); dm != nil {
					if lst, ok := dm["data"].([]any); ok {
						items = lst
					}
				}
				for _, it := range items {
					mm, _ := it.(map[string]any)
					if mm == nil {
						continue
					}
					sid := core.JStr(mm, "sid")
					if sid == "" {
						continue
					}
					if _, ok := rooms[sid]; ok {
						continue
					}
					name := strings.TrimSpace(core.JStr(mm, "name"))
					desc := strings.TrimSpace(core.JStr(mm, "desc"))
					rooms[sid] = meta{
						sid: sid, ssid: core.JStr(mm, "ssid", "sid"),
						title:    titleRe.ReplaceAllString(desc, ""),
						nickname: core.FirstStr(name, desc),
						group:    group,
						avatar:   strings.TrimSpace(core.JStr(mm, "thumb2", "avatar")),
					}
				}
			}
		}
	}
	core.Infof("  [频道] %s: %d 个在播房间", cat, len(rooms))
	return rooms
}

func (Platform) Fetch(sources []core.Source, ctx *core.Ctx) ([]core.Room, error) {
	core.Infof("[yy] %d 个来源", len(sources))
	set := map[string]bool{}
	for _, s := range sources {
		if s.Kind == "category" {
			set[s.Target] = true
		}
	}
	cats := make([]string, 0, len(set))
	for c := range set {
		cats = append(cats, c)
	}
	sort.Strings(cats)
	type res struct {
		m map[string]meta
	}
	results := core.Parallel(3, cats, func(c string) map[string]meta { return fetchCategory(c) })
	merged := map[string]meta{}
	for _, m := range results {
		for sid, v := range m {
			if _, ok := merged[sid]; !ok {
				merged[sid] = v
			}
		}
	}
	core.Infof("  [列表] 共 %d 个不重复房间", len(merged))
	out := make([]core.Room, 0, len(merged))
	for sid, m := range merged {
		g := m.group
		if g == "" {
			g = "yy"
		}
		out = append(out, core.Room{
			Platform: "yy", RID: sid, Title: m.title, Nickname: m.nickname,
			URL: fmt.Sprintf(playerBase, sid), Group: g, Avatar: m.avatar,
		})
	}
	core.Infof("[yy] 完成: %d 在播房间（地址走 Worker 动态解析，未逐个取流）", len(out))
	return out, nil
}

var _ platform.Platform = Platform{}
